src/box/iproto.cc

```diff
@@ -557,6 +615,110 @@ struct iproto_connection
 	struct iproto_thread *iproto_thread;
 };
 
+static struct fiber_cond alive_connection_exists;
+static struct fiber_cond alive_request_exists;
+
+void
+iproto_wait_close_all_connections(void)
+{
+	if (iproto_connection_count() > 0)
+		fiber_cond_wait(&alive_connection_exists);
+}
+
+void
+check_no_alive_connections(struct iproto_thread *iproto_thread);
+
+void
+tx_check_no_alive_connections(struct cmsg *msg)
+{
+	(void) msg;
+	if (iproto_connection_count() == 0)
+		fiber_cond_broadcast(&alive_connection_exists);
+}
+
+void
+net_check_no_alive_connections(struct cmsg *msg)
+{
+	struct iproto_thread *iproto_thread =
+		container_of(msg, struct iproto_thread,
+			     check_alive_connections_msg);
+	if (iproto_thread->check_alive_connections_status == CHECK_REQUEST) {
+		iproto_thread->check_alive_connections_status =
+			CHECK_NOT_STARTED;
+		check_no_alive_connections(iproto_thread);
+	} else {
+		iproto_thread->check_alive_connections_status =
+			CHECK_NOT_STARTED;
+	}
+}
+
+void
+check_no_alive_connections(struct iproto_thread *iproto_thread)
+{
+	if (mempool_count(&iproto_thread->iproto_connection_pool) > 0)
+		return;
+	int check_status = iproto_thread->check_alive_connections_status;
+	if (check_status == CHECK_PROCESS || check_status == CHECK_REQUEST) {
+		iproto_thread->check_alive_connections_status =
+			CHECK_REQUEST;
+		return;
+	}
+	iproto_thread->check_alive_connections_status = CHECK_PROCESS;
+	cmsg_init(&iproto_thread->check_alive_connections_msg,
+		  iproto_thread->check_connections_route);
+	cpipe_push(&iproto_thread->tx_pipe,
```

Comment:
When you push message without flush pipe it's not sends to other thread immediately.

Reply:
Added flushes